### PR TITLE
Overlapping boundaries

### PR DESCRIPTION
**Pull request by Vatyx**
#### Description
Enables overlapping boundaries for multiple workers. There are still bugs but it's at parity with bugs without overlapping boundaries.

Still cleaning up the PR but it can be reviewed now.

#### Primary reviewers
@improbable-valentyn @girayimprobable

### CODE REVIEW

File: SpatialGDK/Source/Private/EngineClasses/SpatialActorChannel.cpp [hunk 1]
```diff
@@ -200,6 +200,11 @@ bool USpatialActorChannel::ReplicateActor()
 	check(!Closing);
 	check(Connection);
 	check(Connection->PackageMap);
+
+	if (Actor->Role != ROLE_Authority)
```
Comment (m-samiec) [inline SpatialGDK/Source/Private/EngineClasses/SpatialActorChannel.cpp]: Should we move this into IsReadyForReplication? Requires moving the Actor check also.

File: SpatialGDK/Source/Private/EngineClasses/SpatialActorChannel.cpp [hunk 2]
```diff
@@ -562,6 +571,9 @@ void USpatialActorChannel::OnReserveEntityIdResponse(const Worker_ReserveEntityI
   
 	EntityId = Op.entity_id;
 	RegisterEntityId(EntityId);
+
+	// Register Actor with package map since we know what the entity id is.
+	NetDriver->PackageMap ->ResolveEntityActor(Actor, EntityId, improbable::CreateOffsetMapFromActor(Actor));
```
Comment (m-samiec) [inline SpatialGDK/Source/Private/EngineClasses/SpatialActorChannel.cpp]: This is a different PR right?

File: SpatialGDK/Source/Public/EngineClasses/SpatialActorChannel.h [hunk 3]
```diff
@@ -54,6 +54,16 @@ class SPATIALGDK_API USpatialActorChannel : public UActorChannel
 		return NetDriver->View->GetAuthority(EntityId, Info->RPCComponents[RPC_Client]) == WORKER_AUTHORITY_AUTHORITATIVE;
 	}
 
+	FORCEINLINE bool IsAuthoritativeServer()
+	{
+		if (!NetDriver->IsServer())
```
Comment (m-samiec) [inline SpatialGDK/Source/Public/EngineClasses/SpatialActorChannel.h]: nit: avoid the branch -
return NetDriver->IsServer() && ...

File: SpatialGDK/Source/Private/EngineClasses/SpatialActorChannel.cpp [hunk 4]
```diff
@@ -673,11 +674,14 @@ void USpatialActorChannel::SpatialViewTick()
 		{
 			if (NetDriver->IsServer())
 			{
-				bool bSuccess = Sender->UpdateEntityACLs(Actor, GetEntityId());
-
-				if (bFirstTick && bSuccess)
+				if (IsAuthoritativeServer())
```
Comment (m-samiec) [inline SpatialGDK/Source/Private/EngineClasses/SpatialActorChannel.cpp]: Remove the IsServer check above.

File: SpatialGDK/Source/Private/Interop/GlobalStateManager.cpp [hunk 5]
```diff
@@ -60,6 +60,9 @@ void UGlobalStateManager::LinkExistingSingletonActors()
 		USpatialActorChannel* Channel = nullptr;
 		GetSingletonActorAndChannel(Pair.Key, SingletonActor, Channel);
 
+		SingletonActor->Role = ROLE_SimulatedProxy;
```
Comment (m-samiec) [inline SpatialGDK/Source/Private/Interop/GlobalStateManager.cpp]: Shouldn't this be done after the the preceding Channel check?

File: SpatialGDK/Source/Private/Interop/SpatialReceiver.cpp [hunk 6]
```diff
@@ -157,10 +161,67 @@ void USpatialReceiver::OnRemoveEntity(Worker_RemoveEntityOp& Op)
...
+			{
+				if (Op.authority == WORKER_AUTHORITY_AUTHORITATIVE)
+				{
+					Actor->Role = ROLE_Authority;
+					if (Actor->GetNetConnection() != nullptr)
+					{
+						Actor->RemoteRole = ROLE_AutonomousProxy;
+					}
+					// TODO: Change to see it has an APlayerController
+					else if (Cast<APawn>(Actor) != nullptr)
```
Comment (m-samiec) [inline SpatialGDK/Source/Private/Interop/SpatialReceiver.cpp]: Replace with IsA check

File: SpatialGDK/Source/Private/EngineClasses/SpatialActorChannel.cpp (hunk 2 above)
Comment (Vatyx) [reply SpatialGDK/Source/Private/EngineClasses/SpatialActorChannel.cpp]: Yes, but that PR was merged into this one.

File: SpatialGDK/Source/Private/Interop/SpatialReceiver.cpp [hunk 7]
```diff
@@ -157,10 +161,67 @@ void USpatialReceiver::OnRemoveEntity(Worker_RemoveEntityOp& Op)
...
+				}
+			}
+		}
+	}
+	else
 	{
-		GlobalStateManager->ExecuteInitialSingletonActorReplication();
+		if (AActor* Actor = NetDriver->GetEntityRegistry()->GetActorFromEntityId(Op.entity_id))
+		{
+			FClassInfo* Info = TypebindingManager->FindClassInfoByClass(Actor->GetClass());
```
Comment (m-samiec) [inline SpatialGDK/Source/Private/Interop/SpatialReceiver.cpp]: Comment about the functionality here would be useful.

File: SpatialGDK/Source/Private/Interop/SpatialReceiver.cpp [hunk 8]
```diff
@@ -276,8 +346,22 @@ void USpatialReceiver::ReceiveActor(Worker_EntityId EntityId)
 			}
 		}
 
+		// Assume SimulatedProxy until we're delegated Authority
```
Comment (m-samiec) [inline SpatialGDK/Source/Private/Interop/SpatialReceiver.cpp]: Delete code

File: SpatialGDK/Source/Private/Interop/SpatialReceiver.cpp [hunk 9]
```diff
@@ -334,6 +418,11 @@ void USpatialReceiver::RemoveActor(Worker_EntityId EntityId)
 		}
 	}
 
+	if (Actor->GetClass()->HasAnySpatialClassFlags(SPATIALCLASS_Singleton))
```
Comment (m-samiec) [inline SpatialGDK/Source/Private/Interop/SpatialReceiver.cpp]: I think we should replace this with a check

File: SpatialGDK/Source/Private/Interop/SpatialReceiver.cpp [hunk 10]
```diff
@@ -519,7 +608,12 @@ void USpatialReceiver::OnCommandRequest(Worker_CommandRequestOp& Op)
 	if (Op.request.component_id == SpatialConstants::PLAYER_SPAWNER_COMPONENT_ID && CommandIndex == 1)
 	{
 		Schema_Object* Payload = Schema_GetCommandRequestObject(Op.request.schema_type);
-		NetDriver->PlayerSpawner->ReceivePlayerSpawnRequest(GetStringFromSchema(Payload, 1), Op.caller_worker_id, Op.request_id);
+
```
Comment (m-samiec) [inline SpatialGDK/Source/Private/Interop/SpatialReceiver.cpp]: nit: kinda gross, but we should make the command index a spatial constant.

File: SpatialGDK/Source/Private/Interop/SpatialSender.cpp [hunk 11]
```diff
@@ -53,13 +53,19 @@ void USpatialSender::Init(USpatialNetDriver* InNetDriver)
...
 
-Worker_RequestId USpatialSender::CreateEntity(const FString& ClientWorkerId, const FString& EntityType, USpatialActorChannel* Channel)
+Worker_RequestId USpatialSender::CreateEntity(USpatialActorChannel* Channel)
 {
 	AActor* Actor = Channel->Actor;
 
+	FString ClientWorkerAttribute;
+	if (UNetConnection* Connection = Channel->Actor->GetNetConnection())
+	{
+		ClientWorkerAttribute= Connection->PlayerController->PlayerState->UniqueId.ToString();
```
Comment (m-samiec) [inline SpatialGDK/Source/Private/Interop/SpatialSender.cpp]: nit: space

File: SpatialGDK/Source/Private/Interop/SpatialSender.cpp [hunk 12]
```diff
@@ -112,11 +118,11 @@ Worker_RequestId USpatialSender::CreateEntity(const FString& ClientWorkerId, con
 
 	TArray<Worker_ComponentData> ComponentDatas;
 	ComponentDatas.Add(improbable::Position(improbable::Coordinates::FromFVector(Channel->GetActorSpatialPosition(Actor))).CreatePositionData());
-	ComponentDatas.Add(improbable::Metadata(EntityType).CreateMetadataData());
+	ComponentDatas.Add(improbable::Metadata(Channel->Actor->GetClass()->GetPathName()).CreateMetadataData());
```
Comment (m-samiec) [inline SpatialGDK/Source/Private/Interop/SpatialSender.cpp]: Why the change?

File: SpatialGDK/Source/Public/Interop/SpatialPlayerSpawner.h [hunk 13]
```diff
@@ -23,7 +23,7 @@ class SPATIALGDK_API USpatialPlayerSpawner : public UObject
 	void Init(USpatialNetDriver* NetDriver, FTimerManager* TimerManager);
 
 	// Server
-	void ReceivePlayerSpawnRequest(FString URLString, const char* CallerWorkerId, Worker_RequestId RequestId);
+	void ReceivePlayerSpawnRequest(FString URLString, const char* CallerWorker, Worker_RequestId RequestId);
```
Comment (m-samiec) [inline SpatialGDK/Source/Public/Interop/SpatialPlayerSpawner.h]: CallerAttribute

**Comment by davedolben** [comment]
GitHub won't let me comment on a line that wasn't modified for some reason, but you need to call `PostNetInit` at the bottom of `USpatialReceiver::ReceiveActor` even if you're on the server, so the deferred BeginPlay actually gets executed. Note that this will crash, since `PostNetInit` has a check for authority, so you'll probably just have to copy-paste its contents.

File: SpatialGDK/Source/Private/EngineClasses/SpatialActorChannel.cpp [hunk 14]
```diff
@@ -576,10 +588,19 @@ void USpatialActorChannel::OnCreateEntityResponse(const Worker_CreateEntityRespo
...
-		return;
+		if (Op.status_code == WORKER_STATUS_CODE_TIMEOUT)
+		{
+			UE_LOG(LogSpatialActorChannel, Error, TEXT("Failed to create entity for actor %s Reason: %s. Retrying..."), *Actor->GetName(), UTF8_TO_TCHAR(Op.message));
+			Sender->SendCreateEntityRequest(this);
+			return;
+		}
+		else
+		{
+			UE_LOG(LogSpatialActorChannel, Error, TEXT("Failed to create entity for actor %s Reason: %s"), *Actor->GetName(), UTF8_TO_TCHAR(Op.message));
```
Comment (danielimprobable) [inline SpatialGDK/Source/Private/EngineClasses/SpatialActorChannel.cpp]: Is this a driveby fix or is it related to the overlapping boundaries work?

File: SpatialGDK/Source/Private/Interop/SpatialReceiver.cpp [hunk 15]
```diff
@@ -157,10 +161,67 @@ void USpatialReceiver::OnRemoveEntity(Worker_RemoveEntityOp& Op)
...
+			if (AActor* Actor = NetDriver->GetEntityRegistry()->GetActorFromEntityId(Op.entity_id))
+			{
+				if (Op.authority == WORKER_AUTHORITY_AUTHORITATIVE)
+				{
+					Actor->Role = ROLE_Authority;
+					if (Actor->GetNetConnection() != nullptr)
+					{
+						Actor->RemoteRole = ROLE_AutonomousProxy;
+					}
+					// TODO: Change to see it has an APlayerController
```
Comment (danielimprobable) [inline SpatialGDK/Source/Private/Interop/SpatialReceiver.cpp]: Add Jira ticket if you don't intend to fix this with the PR

File: SpatialGDK/Source/Private/Interop/SpatialSender.cpp (hunk 12 above)
Comment (danielimprobable) [reply SpatialGDK/Source/Private/Interop/SpatialSender.cpp]: Do we need null checks here?

Comment (Vatyx) [reply SpatialGDK/Source/Private/Interop/SpatialSender.cpp]: Change is because `EntityType` was being passed as a parameter when it didn't need to be. No null checks are needed since you have the actor at this point in every case.

File: SpatialGDK/Source/Private/Interop/SpatialReceiver.cpp [hunk 16]
```diff
@@ -157,10 +161,67 @@ void USpatialReceiver::OnRemoveEntity(Worker_RemoveEntityOp& Op)
...
+	}
+
+	HandleActorAuthority(Op);
+}
+
+void USpatialReceiver::HandleActorAuthority(Worker_AuthorityChangeOp& Op)
+{
+	if (NetDriver->IsServer())
+	{
+		if (Op.component_id == SpatialConstants::GLOBAL_STATE_MANAGER_COMPONENT_ID
```
[the remark this answers is not in the document]
Comment (Vatyx) [reply SpatialGDK/Source/Private/Interop/SpatialReceiver.cpp]: This handles the case that the GSM can go through an authority transition.

File: SpatialGDK/Source/Private/Interop/SpatialReceiver.cpp (hunk 15 above)
Comment (Vatyx) [reply SpatialGDK/Source/Private/Interop/SpatialReceiver.cpp]: I fixed it now rather than adding a ticket.

File: SpatialGDK/Source/Private/EngineClasses/SpatialActorChannel.cpp (hunk 14 above)
Comment (Vatyx) [reply SpatialGDK/Source/Private/EngineClasses/SpatialActorChannel.cpp]: Yes it is. Most likely will take it out if James's changes go in.

File: SpatialGDK/Source/Private/Interop/SpatialSender.cpp [hunk 17]
```diff
@@ -441,7 +450,10 @@ void USpatialSender::QueueOutgoingUpdate(USpatialActorChannel* DependentChannel,
 	*Unresolved = UnresolvedObjects;
 
 	FHandleToUnresolved& HandleToUnresolved = PropertyToUnresolved.FindOrAdd(ChannelObjectPair);
-	check(!HandleToUnresolved.Find(Handle));
+	if (HandleToUnresolved.Find(Handle))
```
Comment (girayimprobable) [inline SpatialGDK/Source/Private/Interop/SpatialSender.cpp]: Can you elaborate on what has changed here so that we don't need to assert?

Also, why do you not simply skip rather than removing and adding?

Comment (girayimprobable) [reply SpatialGDK/Source/Private/Interop/SpatialSender.cpp]: Hmm I guess the unresolved UObject pointer can change before we resolve.

File: SpatialGDK/Source/Private/Interop/SpatialSender.cpp [hunk 18]
```diff
@@ -646,13 +658,30 @@ bool USpatialSender::UpdateEntityACLs(AActor* Actor, Worker_EntityId EntityId)
 	FClassInfo* Info = TypebindingManager->FindClassInfoByClass(Actor->GetClass());
 	check(Info);
 
-	FString PlayerWorkerId;
+	FString PlayerWorkerAttribute;
 	if (Actor->GetNetConnection() != nullptr)
 	{
-		PlayerWorkerId = Actor->GetNetConnection()->PlayerController->PlayerState->UniqueId.ToString();
+		if (APlayerController* PlayerController = Actor->GetNetConnection()->PlayerController)
```
Comment (girayimprobable) [inline SpatialGDK/Source/Private/Interop/SpatialSender.cpp]: Add some commentary to this function please

File: SpatialGDK/Source/Private/Interop/SpatialView.cpp [hunk 19]
```diff
@@ -125,6 +125,19 @@ Worker_Authority USpatialView::GetAuthority(Worker_EntityId EntityId, Worker_Com
 	return WORKER_AUTHORITY_NOT_AUTHORITATIVE;
 }
 
+bool USpatialView::HasAuthority(Worker_EntityId EntityId, Worker_ComponentId ComponentId)
+{
+	if (TMap<Worker_ComponentId, Worker_Authority>* ComponentAuthorityMap = EntityComponentAuthorityMap.Find(EntityId))
+	{
+		if (Worker_Authority* Authority = ComponentAuthorityMap->Find(ComponentId))
+		{
+			return *Authority == WORKER_AUTHORITY_AUTHORITATIVE;
```
Comment (girayimprobable) [inline SpatialGDK/Source/Private/Interop/SpatialView.cpp]: Are you intentionally omitting `WORKER_AUTHORITY_AUTHORITY_LOSS_IMMINENT`?

File: SpatialGDK/Source/Private/Interop/SpatialReceiver.cpp [hunk 20]
```diff
@@ -157,10 +161,71 @@ void USpatialReceiver::OnRemoveEntity(Worker_RemoveEntityOp& Op)
...
+			return;
+		}
+
+		// If we became authoritative over the position component. set our role to be ROLE_Authority
+		// and set our RemoteRole to be ROLE_AutonomousProxy iff the actor has an owning connection.
+		if (Op.component_id == SpatialConstants::POSITION_COMPONENT_ID)
+		{
+			if (AActor* Actor = NetDriver->GetEntityRegistry()->GetActorFromEntityId(Op.entity_id))
+			{
+				if (Op.authority == WORKER_AUTHORITY_AUTHORITATIVE)
```
Comment (girayimprobable) [inline SpatialGDK/Source/Private/Interop/SpatialReceiver.cpp]: Not possible to hit this with authority loss imminent?

File: SpatialGDK/Source/Private/Interop/SpatialSender.cpp (hunk 17 above)
Comment (Vatyx) [reply SpatialGDK/Source/Private/Interop/SpatialSender.cpp]: Ah this is because when we try to create an entity again (because of a failure) we need to remove stuff that we've already queued.

File: SpatialGDK/Source/Private/Interop/SpatialReceiver.cpp (hunk 20 above)
Comment (Vatyx) [reply SpatialGDK/Source/Private/Interop/SpatialReceiver.cpp]: We should do another pass when we want to properly enable soft handover.

Comment (girayimprobable) [reply SpatialGDK/Source/Private/Interop/SpatialReceiver.cpp]: I created UNR-640 for you, please add as a comment here and the other place

File: SpatialGDK/Source/Private/Interop/SpatialSender.cpp [hunk 21]
```diff
@@ -634,6 +639,43 @@ void USpatialSender::ResolveOutgoingRPCs(UObject* Object)
 	}
 }
 
+FString USpatialSender::GetOwnerWorkerAttribute(AActor* Actor)
+{
```
Comment (girayimprobable) [inline SpatialGDK/Source/Private/Interop/SpatialSender.cpp]: Add a comment on the purpose of this function and reasoning for various codepaths it can take.

File: SpatialGDK/Source/Private/Interop/SpatialSender.cpp [hunk 22]
```diff
@@ -634,6 +639,43 @@ void USpatialSender::ResolveOutgoingRPCs(UObject* Object)
 	}
 }
 
+FString USpatialSender::GetOwnerWorkerAttribute(AActor* Actor)
```
Comment (girayimprobable) [inline SpatialGDK/Source/Private/Interop/SpatialSender.cpp]: Is there a race condition where owner is still not resolved? In which case we return worker attribute that will be promptly invalidated.

Comment (Vatyx) [reply SpatialGDK/Source/Private/Interop/SpatialSender.cpp]: Potentially? This would be fixed in the next tick though.

File: SpatialGDK/Source/Private/Interop/SpatialSender.cpp (hunk 21 above)
Comment (Vatyx) [reply SpatialGDK/Source/Private/Interop/SpatialSender.cpp]: Added comments.
